pkg/nats/stream/producer_client.go

```diff
@@ -0,0 +1,142 @@
+package stream
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/nats-io/nats.go"
+	"github.com/rs/zerolog/log"
+	"sync"
+	"time"
+)
+
+type ProducerClientParams struct {
+	Conn *nats.Conn
+}
+
+type ProducerClient struct {
+	Conn *nats.Conn
+	mu   sync.RWMutex // Protects access to activeStreamIds and activeConnHeartBeatRequestSubjects
+
+	activeStreamIds                    map[string][]string // A map of ConnId to StreamId that are active
+	activeConnHeartBeatRequestSubjects map[string]string   // A map of ConnId to the subject where a heart beat request should be sent
+	heartBeatCancelFunc                context.CancelFunc
+}
+
+func NewProducerClient(params ProducerClientParams) (*ProducerClient, error) {
+	nc := &ProducerClient{
+		Conn:                               params.Conn,
+		activeStreamIds:                    make(map[string][]string),
+		activeConnHeartBeatRequestSubjects: make(map[string]string),
+	}
+
+	return nc, nil
+}
+
+func (pc *ProducerClient) AddConnDetails(ctx context.Context, connDetails *ConnectionDetails) {
+	pc.mu.Lock()
+	defer pc.mu.Unlock()
+	pc.activeStreamIds[connDetails.ConnId] = append(pc.activeStreamIds[connDetails.ConnId], connDetails.StreamId)
+	pc.activeConnHeartBeatRequestSubjects[connDetails.ConnId] = connDetails.HeartBeatRequestSub
+
+	if pc.heartBeatCancelFunc == nil {
+		go pc.heartBeat(ctx)
+	}
+
+}
+
+func (pc *ProducerClient) RemoveConnDetails(connDetails *ConnectionDetails) {
+	pc.mu.Lock()
+	defer pc.mu.Unlock()
+
+	activeStreamIdsForConn, ok := pc.activeStreamIds[connDetails.ConnId]
+	if !ok {
+		return
+	}
+
+	for i, v := range activeStreamIdsForConn {
+		if v == connDetails.StreamId {
+			activeStreamIdsForConn = append(activeStreamIdsForConn[:i], activeStreamIdsForConn[i+1:]...)
+		}
+	}
+
+	if len(activeStreamIdsForConn) == 0 {
+		delete(pc.activeStreamIds, connDetails.ConnId)
+		delete(pc.activeConnHeartBeatRequestSubjects, connDetails.StreamId)
+	}
+
+	if len(pc.activeStreamIds) == 0 && pc.heartBeatCancelFunc != nil {
+		pc.heartBeatCancelFunc()
+		pc.heartBeatCancelFunc = nil
+	}
+
+}
+
+func (pc *ProducerClient) heartBeat(ctx context.Context) {
+	ctxWithCancel, cancel := context.WithCancel(ctx)
+	pc.heartBeatCancelFunc = cancel
+
+	ticker := time.NewTicker(10 * time.Second)
```

Comment:
Should make it customizabel

Reply:
@wdbaruni  I have customized it. But our configs are really all over place. The best place i see to integrate these are 

NATSConfig, and then from there you pass it on creation of computer_handler on registering of endpoint. Do you have any other place where this will be well suited to integrate ?

Reply:
I agree our configs are all over the place for now. They are going through major refactoring and I wouldn't recommend touching them for now. You can put these as constants in proxy package , or as default values in the stream package.

Reply:
Makes sense. Will do that.